Реквест от 25.12

Dmitriy-prg opened the pull request.

Commit:
Add project files.

Commit:
Правка замечаний

Commit:
Исправление ошибок Swagger

Commit:
Правка замечапний: суффикс "Async", CancellationToken. + Доработки.

AleksandrPopov96 commented on WebApplicationMustToHave/WebApplicationMustToHave.csproj:
```diff
@@ -0,0 +1,27 @@
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <GenerateDocumentationFile>True</GenerateDocumentationFile>
+    <BaseOutputPath></BaseOutputPath>
```
Пустая сигнатура с настройкой - лучше избегать таких вещей, т.к. могут быть причиной различных багов.

Dmitriy-prg replied:
Исправлено

AleksandrPopov96 commented on WebApplicationMustToHave/appsettings.Development.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "Logging": {
```
Дублирование конфигурации. В базовом appsettings у тебя установлены такие же настройки, поэтому здесь секцию можно удалить. Источники конфигурации в .net имеют приоритет, познакомится с ним можно в этой статье https://metanit.com/sharp/aspnet6/6.4.php

Dmitriy-prg replied:
Исправлено

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbBitrate.cs:
```diff
@@ -0,0 +1,56 @@
+    /// </summary>
+    /// <typeparam name="T">Тип значения объема.</typeparam>
+    /// <typeparam name="M">Тип единицы измерения, который должен реализовывать интерфейс IDbMeasureUnit.</typeparam>
+    public interface IDbBitrate<T, M> where M : IDbMeasureUnit
```
Замечания касаются всех моделей, но напишу здесь, т.к. они однотипны:
1) С какой целью задаешь для каждой модели интерфейс? Мое мнение в этом нет никакого смысла, т.к. модель БД чаще всего представляет собой просто DTO объект для мапинга из реляционной таблицы. Там никогда не содержится бизнес логика, максимум, что там есть -  атрибуты для мапинга в реляционную структуру (не рекомендую, используемый тобой подход с конфигурациями куда более гибкий). Поэтому стоит удалить все интерфейсы. Лучше создать интерфейсы/абстрактные классы для каких то базовых вещей, например, у тебя часто встречаются модели у которых есть Id и Name, создав базовый абстрактный класс можно все похожие модели унаследовать от него и каждый раз не писать однотипные поля. Также общая абстракция поможет реализовывать  гибкие подходы получения однотипных данных из БД через дженерики.
2) Рекомендую использовать стандартные типы для полей БД:
 *  long - идентификаторы, внешние ключи
 * double/decimal - числа с плавающей точкой (decimal - если важна точность)
 * int - целочисленные значение, которые точно туда поместятся, например, года у даты
 * enum - для полей которые имеют четко определенное и неизменное количество значений (например, дни недели, либо состояния объекта)
 * bool - логические значения
 * DateTime/DatetimeOffset - для дат без времени/со временем (также для дат со временем рекомендуется хранить их в БД в UTC, чтобы можно было легко получить дату в таймзоне пользователя)
 * string - текстовая информация
Использование стандартных типов позволит избежать возможных проблем с мапингом данных при использовании EF.
3) Для того чтобы не писать конфигурации познакомься с конвенцией наименований в EF https://learn.microsoft.com/en-us/ef/core/modeling/relationships/conventions. Например, поле Id по умолчанию рассматривается как первичный ключ, комбинация NavigationProperty-NavigationPropertyId автоматически конфигурируется как отношение один ко многим и т.п.

Dmitriy-prg replied:
Исправлено

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbComposition.cs:
```diff
@@ -0,0 +1,272 @@
+        /// <summary>
+        /// Ключ битрейта произведения (необязательно).
+        /// </summary>
+        public uint? DbBitrate_BitrateVideoId { get; set; }
```
Нарушение конвенции наименования - свойства принято называть в PascalCase

Dmitriy-prg replied:
Исправлено

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbComposition.cs:
```diff
@@ -0,0 +1,272 @@
+        /// <summary>
+        /// Ключт типа произведения.
+        /// </summary>
+        public uint DbCompositionTypeId { get; set; }
```
1) Зачем ты к части названий полей делаешь приписку Db? Итак ведь понятно что это относится к базе данных
2) Располагай свойства с идентификатором и навигационные свойства рядом чтобы сразу понимать есть ли навигационное свойство или нет, не проматывая всю модель

Dmitriy-prg replied:
1. Поля названы по названию класса. Есть такие же классы без Db (в бизнес модели)
2. Исправлено

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbBitrate.cs:
```diff
@@ -0,0 +1,56 @@
+    /// <summary>
+    /// Представляет собой объект битрейта. Реализует интерфейс IDbBitrate.
+    /// </summary>
+    public class DbBitrate : IDbBitrate<uint, DbMeasureUnit>
```
Какую функцию выполняет дженерики в интерфейсе? Интерфейс тут в принципе не нужен, также у тебя всего одна реализация и дженерик усложняет восприятие кода

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbComposition.cs:
```diff
@@ -0,0 +1,272 @@
+        public DbResolution? Resolution { get; set; }
+
+        /// <summary>
+        /// Возвращает или задает список жанров произведения, с универсальным типом значения T (необязательно).
```
`с универсальным типом значения T` - не понятно заем эта информация в этом и похожих комментариях, если используется конкретный тип. Также в автодокмуентации обычно кратко описывается предназначение свойства, а не операции с ним. В данном случае для понимания хватило бы `Список жанров произведения`

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbGenre.cs:
```diff
@@ -0,0 +1,57 @@
+        /// <summary>
+        /// Возвращает или устанавливает значение жанра.
+        /// </summary>
+        public T Value { get; set; }
```
Еще один очень странный ненужный дженерик - а какой тип поля `Value` кроме строки еще предполагалось использовать для задания названия жанра?

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbMeasureUnit.cs:
```diff
@@ -0,0 +1,47 @@
+        /// <summary>
+        /// Возвращает или задает уникальный идентификатор единицы измерения.
+        /// </summary>
+        /// <remarks>Это свойство является обязательным и не может иметь значения null.</remarks>
```
Лишняя информация, если используешь nullable дерективу в csproj и тип без `?` это итак очевидно. Старайся делать автодокмуентацию максимально краткой и указывать там только то что может быть неявно по коду

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbBitrate.cs:
```diff
@@ -0,0 +1,56 @@
+        /// <summary>
+        /// Возвращает или задает уникальный идентификатор битрейта.
+        /// </summary>
+        public uint Id { get; set; }
```
Если у интерфейса модификатор доступа public, то на свойства его навешивать не нужно - они по умолчанию будут иметь такой же модификатор

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbPerson.cs:
```diff
@@ -0,0 +1,91 @@
+        public uint? YearBirth { get; set; }
+
+        /// <summary>
+        /// Навигационное свойство.
```
Бесполезный комментарий, который не дает никакой информации. Стоит заменить на описание свойства или удалить. Для кейсов ниже сделать аналогично

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbPerson.cs:
```diff
@@ -0,0 +1,91 @@
+        /// <summary>
+        /// Получает строку-представление человека.
+        /// </summary>
+        public string View { get => Name + " " + Surname + " " + Patronymic ?? "" + " " + YearBirth + " г." ?? ""; }
```
Тут сразу несколько проблем:
1) Неинформативное название свойства, без просмотра кода не понятно что тут хранится
2) Если `Patronymic` не будет заполнено, то в строке будет 2 пробела
3) Строка собирается крайне неоптимально на каждую конкатенацию будет создан отдельный объект строки. Чтобы избежать этого используй интерполированные строки (`$"{Name} {Surname}"`) либо `StringBuilder`

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbPerson.cs:
```diff
@@ -0,0 +1,91 @@
+        /// <summary>
+        /// Навигационное свойство.
+        /// </summary>
+        public List<DbComposition>? DbComposition_Performers { get; set; }
```
Нарушение конвенции наименования

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/DbResolution.cs:
```diff
@@ -0,0 +1,45 @@
+        /// <summary>
+        /// Возвращает или устанавливает высоту разрешения.
+        /// </summary>
+        public required uint Height { get; set; }
```
думаю тут с головой хватит и int типа, не зачем задавать такой широкий тип. Аналогично и для ширины

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/EnumsAndConstants.cs:
```diff
@@ -0,0 +1,37 @@
+﻿namespace WebApplicationMustToHave.DataModels
+{
+    ///// <summary>
```
1) Файл не используется - удали
2) Рекомендуется создавать каждый класс/структуру/енам/интерфейс в отдельном файле, если нет иной договоренности на проекте или так делать просто удобнее (например, помещать команду и обработчик медиатора в один файл, чтобы по команде быстро находить обработчик)

AleksandrPopov96 commented on WebApplicationMustToHave/DataModels/ErrorViewModel.cs:
```diff
@@ -0,0 +1,9 @@
+namespace WebApplicationMustToHave.Models
+{
+    public class ErrorViewModel
```
Стоит поместить класс в другое место, т.к. как эта модель не относится к уровне БД, а является частью уровня представления

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/AppDbContext.cs:
```diff
@@ -0,0 +1,105 @@
+
+namespace WebApplicationMustToHave.Repository
+{
+    public interface IAppDbContext
```
Стоит разделять абстракции и реализацию - интерфейс должен быть в отдельном файле. Если говорить про слой работы с БД в целом, то обычно поступают следующим образом:
1) В отдельный проект выделяют все абстракции и модели (интерфейс контекста БД, модели БД, интерфейсы репозиториев и т.п.). Данный проект не зависим от типа БД и используется в качестве "контракта" при работе с любой БД.
2) Под каждый тип БД, с которым может работать приложение создается отдельный проект. В данном проекте располагаются реализации репозиториев, реализация контекста БД и конфигурации моделей под конкретную БД.
Такой подход позволяет отделить конкретную реализацию слоя работы с БД от бизнес логики, повысить тестируемость приложения и при необходимости бесшовно заменить тип БД, переписав только реализацию и изменив регистрацию

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/AppDbContext.cs:
```diff
@@ -0,0 +1,105 @@
+
+        public class AppDbContext: DbContext, IAppDbContext
+    {
+        public DbSet<DbCompositionType> CompositionTypes { get; set; } = null!;
```
писать ` = null!;` у каждого DbSet'а не нужно, EF точно проинициализирует эти свойства

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/AppDbContext.cs:
```diff
@@ -0,0 +1,105 @@
+
+        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
+        {
+            var config = new ConfigurationBuilder()
```
При такой регистрации, если строка подключения будет задана не в appsettings, то упадет исключение. Лучше изменить способ регистрации контекста и использовать подход описанный в разделе DbContext in dependency injection for ASP.NET Core в статье https://learn.microsoft.com/en-us/ef/core/dbcontext-configuration/#dbcontext-in-dependency-injection-for-aspnet-core. Ниже оставляю пример как мы регистрируем контекст БД на одном из проектов

```
services.AddDbContext<IAuthDbContext, OracleAuthDbContext>(options => options
    .UseOracle(connectionString)
    .UseUpperSnakeCaseNamingConvention() // выбираем конвенцию наименования - подробнее расскажу в разделе конвенций
    .EnableSensitiveDataLogging());
```

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/AppDbContext.cs:
```diff
@@ -0,0 +1,105 @@
+        {
+            base.OnModelCreating(modelBuilder);
+
+            modelBuilder.ApplyConfiguration(new DbCompositionTypeConfiguration());
```
Крайне неудачный подход к подключению конфигураций - каждый раз при добавлении новой модели нужно помнить о том что их нужно подключить здесь. Более удачный подход - это сканирование сборки для подключения всех конфигураций. Есть следующий метод `modelBuilder.ApplyConfigurationsFromAssembly(<сборка где искать конфигурации>);`, который позволяет в одну строку подключить все конфигурации моделей

AleksandrPopov96 commented on WebApplicationMustToHave/Program.cs:
```diff
@@ -0,0 +1,110 @@
+}
+else
+{
+    app.UseSwagger();
```
Очень странная логика, что сваггер не доступен в дев среде. Как по мне сваггер должен быть доступен всегда, чтобы понимать, какие методы есть в АПИ и как с ними работать - вынеси подключение сваггера из условия

AleksandrPopov96 commented on WebApplicationMustToHave/Program.cs:
```diff
@@ -0,0 +1,110 @@
+
+app.UseAuthorization();
+
+//app.MapDefaultControllerRoute();
```
Ненужные комментарии и закоментированный код очень захламляют файлы с исходниками. Вместо того чтобы оставлять их - удаляй, при необходимости все изменения можно достать из гита

AleksandrPopov96 commented on WebApplicationMustToHave/Program.cs:
```diff
@@ -0,0 +1,110 @@
+//    }
+//});
+
+app.Run();
```
Отсутствует вызов `app.MapControllers();` перед этой строкой - все твои методы API будут недоступны для прямого вызова

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/DbBitrateConfiguration.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        public void Configure(EntityTypeBuilder<DbBitrate> builder)
+        {
+            builder.ToTable("bitrates");
```
По наличию миграций видно, что ты используешь подход Code First для формирования схемы БД, тогда зачем ты переименовываешь стандартные названия таблиц? В этом нет необходимости. Единственное, что посоветую, ознакомится и использовать пакет https://github.com/efcore/EFCore.NamingConventions, который позволяет использовать различные конвенции наименования при создания таблиц в БД. рекомендую использовать UseSnakeCaseNamingConvention, т.к. такая конвенции привычна для всех СУБД и упростить тебе работу в IDE предназначенных для БД

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/DbBitrateConfiguration.cs:
```diff
@@ -0,0 +1,16 @@
+        public void Configure(EntityTypeBuilder<DbBitrate> builder)
+        {
+            builder.ToTable("bitrates");
+            builder.HasOne<DbComposition>().WithOne(c => c.BitrateVideo).HasForeignKey<DbComposition>(c => c.DbBitrate_BitrateVideoId);
```
Ознакомься с конвенциями наименования EF (скидывал в комментариях ниже), возможно часть конфигураций моделей БД просто станут не нужны  их можно будет удалить.

Кроме того обрати внимание на следующих подход:
1) Для похожих моделей БД можно создать базовую абстракцию
2) Для схожих конфигураций БД можно также создать базовую абстрактную конфигурацию и ограничив её соответствующим дженериком типа из пункта 1
3) Конфигурации конкретных моделей можно унаследовать от базовой абстрактной конфигурации под данную иерархию моделей и не писать часть кода

Такой подход бывает очень полезен когда есть несколько моделей со схожей конфигурацией

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/DbCompositionManager.cs:
```diff
@@ -0,0 +1,246 @@
+    /// <summary>
+    /// Менеджер произведений (интерфейс)
+    /// </summary>
+    public interface IDbCompositionManager
```
Таким подходом - использованием божественного класса для получения данных -  ты лишаешь себя всех преимуществ от использования EF. Кроме того ты добавляешь еще один ненужный слой абстракции - у тебя уже есть интерфейс контекста БД, а ты еще вставляешь свой менеджер перед ним.

Рекомендую сделать следующее - отказаться от этой абстракции и напрямую использовать интерфейс контекста EF и LINQ для написания запросов к БД.

Единственный кейс, когда стоит использовать дополнительные абстракции, это необходимость написания кастомных запросов под конкретную БД. В этом случае нужно создать отдельный интерфейс репозитория и запросы помещать туда.

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/DbCompositionManager.cs:
```diff
@@ -0,0 +1,246 @@
+    /// </summary>
+    public class DbCompositionManager: IDbCompositionManager
+    {
+        private readonly AppDbContext _db;
```
Почему инжектишь реализацию а не интерфейс?

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/DbCompositionManager.cs:
```diff
@@ -0,0 +1,246 @@
+        /// <returns>список</returns>
+        public async Task<List<IViewable>> GetCompositionViewsAsync(CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested) return [];
```
Некорректно используешь токены отмены. Метод `ToListAsync()` принимает токен отмены в качестве параметра стоит передавать его туда и этого будет достаточно для своевременной отмены операции

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/DbCompositionManager.cs:
```diff
@@ -0,0 +1,246 @@
+        public async Task<List<IViewable>> GetCompositionViewsAsync(CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested) return [];
+            if (_db.CompositionTypes == null | _db.CompositionTypes.Local.Count == 0) _db.CompositionTypes?.Load();
```
Условие не нужно, хватит запроса на 105 строке

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/DbCompositionManager.cs:
```diff
@@ -0,0 +1,246 @@
+        {
+            if (cancellationToken.IsCancellationRequested) return [];
+            if (_db.CompositionTypes == null | _db.CompositionTypes.Local.Count == 0) _db.CompositionTypes?.Load();
+            List<IViewable> compositions = await _db.Compositions.Where(c => c != null).Select(c => (c as IViewable)!).ToListAsync();
```
1) Кастинг в интерфейс не нужен (сам интерфейс также не нужен, об этом писал выше)
2) Проверка на null в Where не нужна, объекты доставаемые из Бд не могут быть null (их поля могут быть, но сами объекты нет)
3) Если ты не планируешь изменять свойства объектов и применять изменения в БД, используй `AsNoTracking()` при выполнении запроса. Вызов этого метода может кратно ускорить выполнение запроса. Подробнее читай https://learn.microsoft.com/en-us/ef/core/querying/tracking

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/DbCompositionManager.cs:
```diff
@@ -0,0 +1,246 @@
+            if (cancellationToken.IsCancellationRequested) return [];
+            if (_db.CompositionTypes == null | _db.CompositionTypes.Local.Count == 0) _db.CompositionTypes?.Load();
+            List<IViewable> compositions = await _db.Compositions.Where(c => c != null).Select(c => (c as IViewable)!).ToListAsync();
+            return compositions ?? [];
```
Список точно не будет null ` ?? []` излишне

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/DbCompositionManager.cs:
```diff
@@ -0,0 +1,246 @@
+                _db.Compositions.Update((dbComposition as DbComposition)!);
+                await _db.SaveChangesAsync(cancellationToken);
+            }
+            else Console.WriteLine("!!!Warning UpdateCompositionAsync() dbComposition = null");
```
1) инвертируй условие на 172 строке и вместо записи в консоль либо используй ILogger для логирования ошибки, а лучше выброси исключение, т.к. при нормальной работе приложения такая ситуация невозможна. Вывод сообщения в консоль не спасет тебя вне локальной отладки

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/DbCompositionManager.cs:
```diff
@@ -0,0 +1,246 @@
+            if (dbComposition != null)
+            {
+                if (cancellationToken.IsCancellationRequested) return;
+                _db.Compositions.Remove(dbComposition);
```
начиная с 6 .net в EF есть куда более оптимальный подход к удалению и обновлении объектов. Ознакомься с https://learn.microsoft.com/en-us/ef/core/saving/execute-insert-update-delete и используй. Текущая реализация синхронно сходит в БД за сущностью, чтобы потом её удалить - как минимум 1 лишний запрос, да и вообще это не рационально

AleksandrPopov96 commented on WebApplicationMustToHave/Repository/DbCompositionManager.cs:
```diff
@@ -0,0 +1,246 @@
+                if (cancellationToken.IsCancellationRequested) return;
+                if (dbComposition.Id == 0)
+                {
+                    dbComposition.Id = _db.Compositions.Select(c => c.Id).Max()+1;
```
Так делать категорически нельзя, т.к. если будут конкурентные запросы (2 и более запросов одновременно) то произойдет ошибка уровня БД, т.к. будет вставлена запись с таким же Id, что уже есть в БД. Более того даже если такая ситуация никогда не возникнет, с ростом числа данных данный запрос будет работать все медленнее и медленнее.

Есть 2 варианта решения проблемы:
1) Использование последовательностей (sequence) в БД. Создается специальная сущность-последовательность которая позволяет генерить уникальные ID для вставки значений.
2) Некоторые СУБД поддерживают неявный вызов последовательности если в Id не передавать значение. Провайдеры EF для данных СУБД по умолчанию присваивают первичным ключам такой тип полей (например, в Postgre это тип serial или bigserial). Прочитай про используемую тобой БД и посмотри какой тип полей генерируется для первичных ключей, скорее всего тебе нужно просто оставить ID незаполненным и провайдер сделает все за тебя

AleksandrPopov96 commented on WebApplicationMustToHave/Models/Person.cs:
```diff
@@ -0,0 +1,102 @@
+    /// <summary>
+    /// Представляет человека с идентификатором, именем, фамилией и необязательными полями для указания отчества и года рождения.
+    /// </summary>
+    public interface IPerson
```
Интерфейс для моделей которые используются для вывода так же как и с моделями БД является лишней абстракцией и переусложнением кода, т.к. у тебя всего одна реализация каждого интерфейса. Интерфейс удобно использовать когда у тебя иерархия каких-то объектов либо несколько объектов обладают общими свойствами и тебе хочется написать логику, которая не привязана к определенному типу объектов, но тебе нужно чтобы объект обладал нужными полями. В твоем случае предлагаю отказаться от интерфейсов.

AleksandrPopov96 commented on WebApplicationMustToHave/Models/Person.cs:
```diff
@@ -0,0 +1,102 @@
+        /// </summary>
+        public string View { get => Name + " " + Surname + " " + Patronymic ?? "" + " " + YearBirth + " г." ?? ""; }
+
+        public static IPerson? GetObjFromDb(IDbPerson? dbPerson)
```
Познакомься с библиотекой Automapper (https://docs.automapper.org/en/stable/) она позволяет упростить все рутинные процессы, связанные с мапингом объектов. в особенности обрати внимание на метод ProjectTo, который сразу позволяет доставать из БД нужные объекты а также на концепцию профилей (Profile) для регистрации схем мапинга. После применения автомаппера этот и подобные методы станут не нужны

AleksandrPopov96 commented on WebApplicationMustToHave/Models/CommonInterfaces.cs:
```diff
@@ -0,0 +1,38 @@
+﻿using WebApplicationMustToHave.DataModels;
+using WebApplicationMustToHave.Repository;
+
+namespace WebApplicationMustToHave.Models
```
стоит размещать каждый класс в отдельном файле, это упрощает поиск объектов и делает код более читаемым. Сейчас, если не знать, что в этом файле лежит данный интерфейс, то его очень непросто отыскать

AleksandrPopov96 commented on WebApplicationMustToHave/Models/CommonInterfaces.cs:
```diff
@@ -0,0 +1,38 @@
+        /// <summary>
+        /// Получает строку-представление.
+        /// </summary>
+        public string View { get; }
```
Лучше реализуй метод, а не свойство, например с названием `GetView()`

AleksandrPopov96 commented on WebApplicationMustToHave/Models/CommonInterfaces.cs:
```diff
@@ -0,0 +1,38 @@
+        public string View { get; }
+    }
+
+    public interface IDbLinkable<T> where T : IEntity
```
После использования автомапера интерфейс станет не нужен

AleksandrPopov96 commented on WebApplicationMustToHave/Models/Composition.cs:
```diff
@@ -0,0 +1,139 @@
+    /// <typeparam name="U">Тип единицы измерения, который должен реализовывать IMeasureUnit.</typeparam>
+    /// <typeparam name="V">Тип значения жанра.</typeparam>
+    /// <typeparam name="R">Тип версии рейтинга, который должен реализовывать IRatingVersion.</typeparam>
+    public interface IComposition<T, U, V, R>
```
Очень страшный дженериковый интерфейс - сильно усложняет код без какой-то ощутимой пользы, лучше не используй их без необходимости

AleksandrPopov96 commented on WebApplicationMustToHave/Models/ErrorViewModel.cs:
```diff
@@ -0,0 +1,9 @@
+namespace WebApplicationMustToHave.DataModels
+{
+    public class ErrorViewDataModel
```
модель дублируется в папке с моделями БД - удали оттуда и оставь эту

AleksandrPopov96 commented on WebApplicationMustToHave/Controllers/AppController.cs:
```diff
@@ -0,0 +1,114 @@
+        public AppController(IAppDbContext db, IDbCompositionManager cm)
+        {
+            _db = db;
+            if (_db != null) Console.WriteLine("AppController() AppDbContext ID = " + ((AppDbContext)_db).ContextId);
```
Какую пользу несут эти сообщения?

AleksandrPopov96 commented on WebApplicationMustToHave/Controllers/AppController.cs:
```diff
@@ -0,0 +1,114 @@
+    /// �������� ���������
+    /// </summary>
+    [Route("App")]
+    public class AppController : Controller
```
Контроллер не содержит ниодного используемого метода - стоит удалить

AleksandrPopov96 commented on WebApplicationMustToHave/Controllers/PersonController.cs:
```diff
@@ -0,0 +1,126 @@
+        /// </summary>
+        /// <returns></returns>
+        [HttpGet]
+        [Route("Items")]
```
1) Путь эндпоинта лучше указывать в методе с глаголом. В данном случае будет как-то так `[HttpGet("Items"))]`. отдельный атрибут пути избыточен
2) Отсутствует автодокументация на методах контроллера. В реальном проекте это очень важно, т.к. она отображается в сваггер интерфейсе. Познакомься подробнее с настройками сваггера и как включить там отображения автодокументации. Вот ссылка с библиотекой, которая используется по дефолту для свагера в Апи приложениях https://github.com/domaindrivendev/Swashbuckle.AspNetCore

AleksandrPopov96 commented on WebApplicationMustToHave/Controllers/PersonController.cs:
```diff
@@ -0,0 +1,126 @@
+        [Route("Create")]
+        public ActionResult Create(Person? person)
+        {
+            if (ModelState.IsValid && person != null)
```
На модели Person нет ни одного атрибута валидации модели, поэтому вызов `ModelState.IsValid`. Если хочешь организовать валидацию пользуйся библиотекой https://docs.fluentvalidation.net/en/latest/. Она гораздо более шибкая и удобная чем стандартная валидация моделей в .net

Dmitriy-prg replied:
Исправлено

AleksandrPopov96 commented on WebApplicationMustToHave/Controllers/PersonController.cs:
```diff
@@ -0,0 +1,126 @@
+        {
+            if (ModelState.IsValid && person != null)
+            { 
+                Console.WriteLine("Create" + ControllerContext.HttpContext.Request.Path + " - " + person?.Id + " : " + person?.Name + " : " + person?.YearBirth);
```
Для логирования используй ILogger, а не просто пиши в консоль. В проекте могут быть настроены разные правила и провайдеры логирования кроме консоли

Dmitriy-prg replied:
Исправлено

AleksandrPopov96 commented on WebApplicationMustToHave/Controllers/PersonController.cs:
```diff
@@ -0,0 +1,126 @@
+            if (ModelState.IsValid && person != null)
+            { 
+                Console.WriteLine("Create" + ControllerContext.HttpContext.Request.Path + " - " + person?.Id + " : " + person?.Name + " : " + person?.YearBirth);
+                Task taskAdd = pm.AddPersonAsync((IDbPerson)person, new CancellationTokenSource().Token);
```
Некорректно используешь токен отмены. Стоит включать токен отмены в качестве аргумента метода контроллера. т.е. делать вот так `Create(Person? person, CancellationToken cancellationToken = default)` При инициализации контроллера .net сам заполнит его токеном, который создается в рамках контекста запроса пользователя. Другими словами если запрос в апи отменят, то токен перейдет в сигнальное состояние и асинхронная операция отменится.

Dmitriy-prg replied:
Исправлено

AleksandrPopov96 commented on WebApplicationMustToHave/Controllers/PersonController.cs:
```diff
@@ -0,0 +1,126 @@
+        [Route("Items")]
+        public IActionResult Items()
+        {
+            List<IViewable> views = pm.GetPersonViewsAsync(new CancellationTokenSource().Token)!.Result;
```
Так делать категорически нельзя. Почитай на метаните как работать с асинхронными операциями, а также про ключевые слова async/await https://metanit.com/sharp/tutorial/13.3.php

Dmitriy-prg replied:
Исправлено

AleksandrPopov96 commented on WebApplicationMustToHave/Controllers/PersonController.cs:
```diff
@@ -0,0 +1,126 @@
+        /// <returns></returns>
+        [HttpPost]
+        [Route("Create")]
+        public ActionResult Create(Person? person)
```
Метод не ожидает завершения асинхронной операции на 61 строчке, что очень опасно, т.к. контекст Бд будет уничтожен после завершения обработки запроса и если операция в БД к этому моменту не завершится, то упадет ошибка, о которой ты никогда не узнаешь. Внимательно ознакомься с концепцией асинхронной работы по ссылке из метанита, которую прикреплял выше

Dmitriy-prg replied:
Исправлено

AleksandrPopov96 reviewed (changes requested):
Общие замечания:
1) Не настроено логирование. Воспользуйся библиотекой https://nlog-project.org/ или https://github.com/serilog/serilog для организации логирования
2) Не везде есть проверка на null входных ссылочных аргументов в публичных методах. Возможно появление необработанных исключений из-за пользовательских ошибок
3) Отсутствует централизованная обработка всех исключений в проекте. Сейчас необходимо или втыкать везде try-catch либо при возникновении ошибок приложение будет падать. Посмотри как можно организовать такую обработку исключений - один из лучших вариантов использование мидлвара.

Commit:
Исправлено:
1. Дублирование конфигурации. В базовом appsettings у тебя установлены такие же настройки, поэтому здесь секцию можно удалить.
2. На модели Person нет ни одного атрибута валидации модели, поэтому вызов ModelState.IsValid.
3. Рекомендую использовать стандартные типы для полей БД:
long - идентификаторы, внешние ключи
double/decimal - числа с плавающей точкой (decimal - если важна точность)
int - целочисленные значение, которые точно туда поместятся, например, года у даты
enum - для полей которые имеют четко определенное и неизменное количество значений (например, дни недели, либо состояния объекта)
bool - логические значения
DateTime/DatetimeOffset - для дат без времени/со временем (также для дат со временем рекомендуется хранить их в БД в UTC, чтобы можно было легко получить дату в таймзоне пользователя)
string - текстовая информация
Использование стандартных типов позволит избежать возможных проблем с мапингом данных при использовании EF.

Commit:
Восстановление компилябельности

Commit:
Исправлено:
1. Не настроено логирование. Воспользуйся библиотекой https://nlog-project.org/ или https://github.com/serilog/serilog для организации логирования
2. Не везде есть проверка на null входных ссылочных аргументов в публичных методах. Возможно появление необработанных исключений из-за пользовательских ошибок

Dmitriy-prg commented:
> * Не настроено логирование. Воспользуйся библиотекой https://nlog-project.org/ или https://github.com/serilog/serilog для организации логирования
> * Не везде есть проверка на null входных ссылочных аргументов в публичных методах. Возможно появление необработанных исключений из-за пользовательских ошибок

Исправлено

Dmitriy-prg commented:
> 3\. Отсутствует централизованная обработка всех исключений в проекте. Сейчас необходимо или втыкать везде try-catch либо при возникновении ошибок приложение будет падать. Посмотри как можно организовать такую обработку исключений - один из лучших вариантов использование мидлвара.
Это разве не оно?
https://github.com/Dmitriy-prg/WebApplicationMustToHave/blob/c774b4fff0f09b9fe530b5ff98f8ed3a4e29344c/WebApplicationMustToHave/Program.cs?plain=1#L33C1-L49C2
```// Configure the HTTP request pipeline.
if (app.Environment.IsDevelopment())
{
    app.UseSwagger();
    app.UseSwaggerUI(options => // UseSwaggerUI is called only in Development.
    {
        options.SwaggerEndpoint("/swagger/v1/swagger.json", "v1");
        options.RoutePrefix = string.Empty;
    });
    app.UseDeveloperExceptionPage();
}
else
{
    app.UseExceptionHandler("/Home/Error");
    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
    app.UseHsts();
}```

Commit:
Оптимизация работы с БД.
